cmd/registry-replacer/main.go

```diff
@@ -0,0 +1,210 @@
+				}
+			}()
+			return nil
+		},
+	); err != nil {
+		logrus.WithError(err).Fatal("Failed to operate on ci-operator-config")
+	}
+	wg.Wait()
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		logrus.WithError(err).Fatal("Encountered errors")
+	}
+}
+
+// replacer ensures replace directives are in place. It fetches the files via http because using git
+// en masse easily kills a developer laptop whereas the http calls are cheap and can be parallelized without
+// bounds.
+func replacer(
+	githubFileGetterFactory func(org, repo, branch string) githubFileGetter,
+	writer func(name string, data []byte, mode os.FileMode) error,
+) func(*api.ReleaseBuildConfiguration, *config.Info) error {
+	return func(config *api.ReleaseBuildConfiguration, info *config.Info) error {
+		if !referencesDockerfile(config.Images) {
+			return nil
+		}
+
+		originalConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal config for comparison: %w", err)
+		}
+
+		for idx := range config.Images {
+			foundTags, err := ensureReplacement(&config.Images[idx], githubFileGetterFactory(info.Org, info.Repo, info.Branch))
+			if err != nil {
+				return fmt.Errorf("failed to ensure replacements: %w", err)
+			}
+			for _, foundTag := range foundTags {
+				if config.BaseImages == nil {
+					config.BaseImages = map[string]api.ImageStreamTagReference{}
+				}
+				config.BaseImages[foundTag.String()] = api.ImageStreamTagReference{
+					Namespace: foundTag.org,
+					Name:      foundTag.repo,
+					Tag:       foundTag.tag,
+				}
+			}
+		}
+
+		newConfig, err := yaml.Marshal(config)
+		if err != nil {
+			return fmt.Errorf("failed to marshal new config: %w", err)
+		}
+
+		// Avoid filesystem access if possible
+		if bytes.Equal(originalConfig, newConfig) {
+			return nil
+		}
+
+		if err := writer(info.Filename, newConfig, 0644); err != nil {
+			return fmt.Errorf("faild to write %s: %w", info.Filename, err)
+		}
+
+		return nil
+	}
+}
+
+func referencesDockerfile(images []api.ProjectDirectoryImageBuildStepConfiguration) bool {
+	for _, image := range images {
+		if image.DockerfilePath != "" {
+			return true
+		}
+	}
+
+	return false
+}
+
+var registryRegex = regexp.MustCompile(`registry\.svc\.ci\.openshift\.org\/[^\s]+`)
+
+type orgRepoTag struct{ org, repo, tag string }
+
+func (ort orgRepoTag) String() string {
+	return ort.org + "_" + ort.repo + "_" + ort.tag
+}
+
+func ensureReplacement(image *api.ProjectDirectoryImageBuildStepConfiguration, getter githubFileGetter) ([]orgRepoTag, error) {
+	if image.DockerfilePath == "" {
+		return nil, nil
+	}
+
+	data, err := getter(image.DockerfilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get dockerfile %s: %w", image.DockerfilePath, err)
+	}
+
+	var toReplace []string
+	for _, line := range bytes.Split(data, []byte("\n")) {
+		if !bytes.Contains(line, []byte("FROM")) {
+			continue
+		}
+		if !bytes.Contains(line, []byte("registry.svc.ci.openshift.org")) {
```

Comment:
Is this an optimization to not use regex?

Reply:
Yes